data/sqlutil/scanrow.go

```diff
@@ -76,41 +74,53 @@ func MakeScanRow(colTypes []*sql.ColumnType, colNames []string, converters ...Co
 			nullable = true // If we don't know if it is nullable, assume it is
 		}
 
-		var converter *Converter
-		scanType := colType.ScanType()
-		for i, v := range converters {
-			if v.InputTypeRegex != nil {
-				if v.InputTypeRegex.MatchString(colType.DatabaseTypeName()) {
-					scanType = v.InputScanType
-					converter = &converters[i]
-					break
-				}
-			}
-
-			if v.InputColumnName == colName && v.InputColumnName != "" {
-				scanType = v.InputScanType
-				converter = &converters[i]
-				break
-			}
-
-			// If there's an applicable converter for this column, scan using the InputScanType.
-			if v.InputTypeName == colType.DatabaseTypeName() {
-				scanType = v.InputScanType
-				converter = &converters[i]
+		for _, v := range converters {
+			if m := match(v, colType, colName); m {
+				rc.append(colName, scanType(v, colType.ScanType()), v)
 				break
 			}
 		}
 
-		if converter == nil {
-			v := NewDefaultConverter(colType.Name(), nullable, scanType)
-			converter = &v
-			scanType = v.InputScanType
+		if !rc.hasConverter(i) {
+			v := NewDefaultConverter(colType.Name(), nullable, colType.ScanType())
+			rc.append(colName, scanType(v, colType.ScanType()), v)
 		}
-		converter.colType = *colType
+	}
 
-		r.Append(colName, scanType)
-		c = append(c, *converter)
+	return rc, nil
+}
+
+func scanType(v Converter, t reflect.Type) reflect.Type {
+	if v.InputScanType != nil {
+		return v.InputScanType
 	}
+	return t
+}
+
+type RowConverter struct {
+	Row        *ScanRow
+	Converters []Converter
+}
+
+func NewRowConverter() *RowConverter {
+	return &RowConverter{Row: NewScanRow(0)}
+}
+
+func (r *RowConverter) append(name string, kind reflect.Type, conv Converter) {
+	r.Row.Append(name, kind)
+	r.Converters = append(r.Converters, conv)
+}
+
+func (r *RowConverter) hasConverter(i int) bool {
+	return len(r.Converters) >= i
+}
+
+func (r *RowConverter) NewScannableRow() []any {
+	return r.Row.NewScannableRow()
+}
 
-	return r, c, nil
+func match(v Converter, colType *sql.ColumnType, colName string) bool {
+	return (v.InputTypeRegex != nil && v.InputTypeRegex.MatchString(colType.DatabaseTypeName())) ||
+		(v.InputColumnName == colName && v.InputColumnName != "") ||
+		v.InputTypeName == colType.DatabaseTypeName()
```

Comment:
Does it make sense to swap the conditions so that regex is the latest? Will it increase performance in some cases?

Reply:
yes.  updated